sopel/irc/capabilities.py

```diff
@@ -0,0 +1,274 @@
+            params,
+            name in self._available,
+            name in self._enabled,
+        )
+
+    @property
+    def available(self) -> Dict[str, Optional[str]]:
+        """Dict of available server capabilities.
+
+        Each key is the name of a capability advertised by the server, and each
+        value is the parameters as advertised (if any) for this capability.
+
+        If a capability is not in this ``dict``, it means the server doesn't
+        advertise it, and it cannot be requested.
+        """
+        return dict(self._available.items())  # return a copy
+
+    @property
+    def enabled(self) -> FrozenSet[str]:
+        """Set of enabled server capabilities.
+
+        Each element is the name of a capability that is enabled on the server.
+        """
+        return frozenset(self._enabled)
+
+    def is_available(self, name: str) -> bool:
+        """Tell if the capability ``name`` is available on the server."""
+        return name in self._available
+
+    def is_enabled(self, name: str) -> bool:
+        """Tell if the capability ``name`` is enabled on the server."""
+        return name in self._enabled
+
+    def handle_ls(self, bot: SopelWrapper, trigger: Trigger) -> bool:
+        """Handle a ``CAP LS`` command.
+
+        This method behaves as a plugin callable with its ``bot`` and
+        ``trigger`` arguments, with the precise goals to handle ``CAP LS``
+        command only, registering available capabilities.
+
+        Then it returns if there is no more ``LS`` command to handle (in case
+        of multi-line ``LS``).
+        """
+        # dev checking
+        assert trigger.event == 'CAP'
+        assert trigger.args[1] == 'LS'
+
+        # extracting capabilities
+        for available_capability in trigger.split():
+            name, *params = available_capability.split('=', maxsplit=1)
+            self._available[name] = params[0] if params else None
+
+        # return if multiline or not
+        multiline = trigger.args[2] == '*'
+        return not multiline
+
+    def handle_ack(
+        self,
+        bot: SopelWrapper,
+        trigger: Trigger,
+    ) -> Tuple[str, ...]:
+        """Handle a ``CAP ACK`` command.
+
+        This method behaves as a plugin callable with its ``bot`` and
+        ``trigger`` arguments, with the precise goals to handle ``CAP ACK``
+        command only, registering enabled or disabled capabilities and tracking
+        acknowledgement.
+
+        Then it returns the list of acknowledged capability requests.
+        """
+        # dev checking
+        assert trigger.event == 'CAP'
+        assert trigger.args[1] == 'ACK'
+
+        # extracting capabilities
+        ack_capabilities = tuple(sorted(trigger.split()))
+
+        # update enabled capabilities
+        for ack_capability in ack_capabilities:
+            if ack_capability.startswith('-'):
+                # ACK a disable request
+                cap_reversed = ack_capability.strip('-')
```

Comment:
Nitpick: as I understand it, it's possible for a particularly antisocial capability name to have a trailing dash, i.e. `nuisance-cap-`, which would be stripped away here. I'm not sure if a capability is allowed to have a leading dash, though? If so, we could slice here, but I think `lstrip()` probably satisfies my nitpick.

```suggestion
                cap_reversed = ack_capability.lstrip('-')
```

Reply:
Ooh, that's a _very_ edge case. Is this worth bringing up in the #ircv3 channel or issue tracker? I went looking in [the spec](https://ircv3.net/specs/extensions/capability-negotiation) for a formal grammar or allowable charset, but found none; only that capability names MUST be treated opaque identifiers, and that names using lowercase-with-hyphens-between-words is merely a SHOULD.

Reply:
To tie this up once and for all, I proposed ircv3/ircv3-specifications#519, which was accepted. We needn't worry about a leading hyphen!